core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java

```diff
@@ -585,22 +547,24 @@ public static AuthSession fromAccessToken(
       long startTimeMillis = System.currentTimeMillis();
       Long expiresAtMillis = session.expiresAtMillis();
 
-      if (isExpired(expiresAtMillis, startTimeMillis)) {
+      if (null != expiresAtMillis && expiresAtMillis <= startTimeMillis) {
         Pair<Integer, TimeUnit> expiration = session.refresh(client);
+        // if expiration is non-null, then token refresh was successful
```

Comment:
I think we need an else case that avoids scheduling token refresh. If the expiration is null then the refresh was not successful and should not be reattempted. But in that case, `expiresAtMillis` has the original expired value and will cause the last check to schedule a token refresh.

Reply:
good catch, that makes sense. I've pushed a fix where we reset the original `expiresAtMillis` when the token refresh failed